stickynotes/stickynotes.c

```diff
@@ -50,978 +50,1055 @@ set_icon_geometry (GdkWindow *window,
-				gchar *font_color_str;
-
-				color_str = (gchar *)xmlGetProp (node, XML_CHAR ("color"));
-				font_color_str = (gchar *)xmlGetProp (node, XML_CHAR ("font_color"));
-
-				if (color_str || font_color_str)
-					stickynote_set_color (note,
-							color_str,
-							font_color_str,
-							TRUE);
-				g_free (color_str);
-				g_free (font_color_str);
-			}
-
-			/* Retrieve and set the font of the note */
-			{
-				gchar *font_str = (gchar *)xmlGetProp (node, XML_CHAR ("font"));
-				if (font_str)
-					stickynote_set_font (note, font_str,
-							TRUE);
-				g_free (font_str);
-			}
-
-			/* Retrieve the workspace */
-			{
-				char *workspace_str;
-
-				workspace_str = (gchar *)xmlGetProp (node, XML_CHAR ("workspace"));
-				if (workspace_str)
-				{
-					note->workspace = atoi (workspace_str);
-					g_free (workspace_str);
-				}
-			}
-
-			/* Retrieve and set (if any) the body contents of the
-			 * note */
-			{
-				gchar *body = (gchar *)xmlNodeListGetString(doc,
-						node->xmlChildrenNode, 1);
-				if (body) {
-					GtkTextBuffer *buffer;
-					GtkTextIter start, end;
-
-					buffer = gtk_text_view_get_buffer(
-						GTK_TEXT_VIEW(note->w_body));
-					gtk_text_buffer_get_bounds(
-							buffer, &start, &end);
-					gtk_text_buffer_insert(buffer,
-							&start, body, -1);
-				}
-				g_free(body);
-			}
-
-			/* Retrieve and set the locked state of the note,
-			 * by default unlocked */
-			{
-				gchar *locked = (gchar *)xmlGetProp(node, XML_CHAR ("locked"));
-				if (locked)
-					stickynote_set_locked(note,
-						!strcmp(locked, "true"));
-				g_free(locked);
-			}
-		}
-
-		node = node->next;
-	}
-
-	tmp1 = new_notes;
-	/*
-	wnck_screen = wnck_screen_get_default ();
-	wnck_screen_force_update (wnck_screen);
-	*/
-
-	while (tmp1)
-	{
-		StickyNote *note = tmp1->data;
-
-		stickynote_set_visible (note, stickynotes->visible);
-		tmp1 = tmp1->next;
-	}
-
-	g_list_free (new_notes);
-	g_list_free (new_nodes);
-
-	xmlFreeDoc(doc);
+    xmlDocPtr doc = NULL;
+    xmlNodePtr root;
+    xmlNodePtr node;
+    /* WnckScreen *wnck_screen; */
+    GList *new_notes, *tmp1;  /* Lists of StickyNote*'s */
+    GList *new_nodes; /* Lists of xmlNodePtr's */
+    int x, y, w, h;
+
+    /* The XML file is $HOME/.config/mate/stickynotes-applet, most probably */
+    {
+        gchar* file = g_build_filename (g_get_user_config_dir (),
+                                        "mate", "stickynotes-applet.xml", NULL);
+
+        if (g_file_test (file, G_FILE_TEST_EXISTS)) {
+            /* load file */
+            doc = xmlParseFile (file);
+        } else {
+            /* old one */
+            g_free (file);
+
+            file = g_build_filename (g_get_home_dir (),
+                                     ".mate2", "stickynotes_applet", NULL);
+
+            if (g_file_test (file, G_FILE_TEST_EXISTS)) {
+                /* load file */
+                doc = xmlParseFile (file);
+            }
+        }
+
+        g_free (file);
+    }
```

Comment:
no need to add this extra scope for saving file name `gchar* file`

Reply:
Better?

Reply:
yes